server/src/main/java/org/opensearch/index/query/TemplateQueryBuilder.java

```diff
@@ -0,0 +1,186 @@
+    public static final String NAME = "template";
+    public static final String queryName = "template";
+    private final Map<String, Object> content;
+
+    /**
+     * Constructs a new TemplateQueryBuilder with the given content.
+     *
+     * @param content The template content as a map.
+     */
+    public TemplateQueryBuilder(Map<String, Object> content) {
+        this.content = content;
+    }
+
+    /**
+     * Creates a TemplateQueryBuilder from XContent.
+     *
+     * @param parser The XContentParser to read from.
+     * @return A new TemplateQueryBuilder instance.
+     * @throws IOException If there's an error parsing the content.
+     */
+    public static TemplateQueryBuilder fromXContent(XContentParser parser) throws IOException {
+        return new TemplateQueryBuilder(parser.map());
+    }
+
+    /**
+     * Constructs a TemplateQueryBuilder from a stream input.
+     *
+     * @param in The StreamInput to read from.
+     * @throws IOException If there's an error reading from the stream.
+     */
+    public TemplateQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.content = in.readMap();
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeMap(content);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.field(NAME, content);
+    }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        throw new IllegalStateException(
+            "Template queries cannot be converted directly to a query. Template Query must be rewritten first during doRewrite."
+        );
+    }
+
+    @Override
+    protected boolean doEquals(TemplateQueryBuilder other) {
+        return Objects.equals(this.content, other.content);
+    }
+
+    @Override
+    protected int doHashCode() {
+        return Objects.hash(content);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    /**
+     * Gets the content of this template query.
+     *
+     * @return The template content as a map.
+     */
+    public Map<String, Object> getContent() {
+        return content;
+    }
+
+    /**
+     * Rewrites the template query by substituting variables from the context.
+     *
+     * @param queryCoordinatorContext The context for query rewriting.
+     * @return A rewritten QueryBuilder.
+     * @throws IOException If there's an error during rewriting.
+     */
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryCoordinatorContext) throws IOException {
+        // the queryRewrite is expected at QueryCoordinator level
+        if (!(queryCoordinatorContext instanceof QueryCoordinatorContext)) {
+            throw new IllegalStateException(
+                "Template Query must be rewritten at the coordinator node. Rewriting at shard level is not supported."
```

Comment:
Should "not supported" be an `UnsupportedOperationException`?  What is the value of making it an `IllegalStateException`?

Reply:
umm maybe I can make the message more clear, but I think this is the right type of Exception, because [IllegalStateException](https://docs.oracle.com/javase/8/docs/api/java/lang/IllegalStateException.html) is used for `a method has been invoked at an illegal or inappropriate time`. And this is the proper case that ideally, it will never have this exception unless OpenSearch has bug in it.